File: Python/150.Evaluate Reverse Polish Notation.py
#计算逆波兰表达式值,有效运算符为+-*/
class Solution(object):
    def evalRPN(self, tokens):
        stack = []
        for ch in tokens:
            if ch=="+":
                a = stack.pop()
                b = stack.pop()
                stack.append(a+b)
            elif ch=="-":
                a = stack.pop()
                b = stack.pop()
                stack.append(b-a)
            elif ch=="*":
                a = stack.pop()
                b = stack.pop()
                stack.append(a*b)
            elif ch=="/":
                a = stack.pop()
                b = stack.pop()
                flag = 1
                if a<0:
                    a = -a
                    flag = flag * -1
                if b<0:
                    b = -b
                    flag = flag * -1
                stack.append(flag*int(b/a))        #6/-132结果为-1，而答案要求0，所以"//"操作符不好使
            else:
                stack.append(int(ch))
        return stack[0]

s = Solution()
print(s.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))

File: C++/Stack/150.Evaluate Reverse Polish Notation.cpp
/*
给出逆波兰表达式，计算结果值。
*/
/*
思路：如果是数字，进栈。如果是计算符，出栈两次，计算结果，然后压栈。
      判断是否是数字可以判断字符串的最后一个字符的ASCII码是否在0-9之间。
      不能使用第一个是因为会有负数。
*/
#include <iostream>
#include <vector>
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int>	mystack;
        int x,y;
        for(auto s:tokens)
        {
        	//注意这个判断不能用s[0]，因为可能有负数。
        	if(s.back()<'0' || s.back()>'9')
        	{
        		x = mystack.top();
        		mystack.pop();
        		y = mystack.top();
        		mystack.pop();
        		switch(s[0])
        		{
        			case '+':	y = x+y;	break;
        			case '-':	y = y-x;	break;
        			case '*':	y = x*y;	break;
        			case '/':	y = y/x;	break;
        		}
        		mystack.push(y);
        	}
        	else
        		mystack.push(stoi(s));
        }
        return mystack.top();
    }
};
int main()
{
	Solution s;
	vector<string> tokens{"4", "13", "5", "/", "+"};
	cout << s.evalRPN(tokens) <<endl;
}

